### DISCUSSION TITLE
Is formGraph designed as asynchronized?

### DISCUSSION THREAD

**Discussion by RihuaChen** [alibaba/formily#3829] [Q&A 2.x]
1. When I use `form.setFieldState` to set an array value, like

` form.setFieldState('arr', state => {
      state.value = [{arrA: 111, arrB: 222}]
})`

2. then I use `form.getFormGraph()` function immediately,  I will not get 

<img width="357" alt="image" src="https://github.com/alibaba/formily/assets/11918443/7b2058cc-bf18-4be4-932b-a4f3eab0d28d">

3. if i use setTimeout,
`setTimeout(() => {
   console.log(form.getFormGraph())
}, [])
`
I will get the final correct FormGraph.
<img width="344" alt="image" src="https://github.com/alibaba/formily/assets/11918443/10756fe4-b522-4fb8-8fee-f2decb10e564">

can I get formGraph synchronized, or it is designed as asynchronized?

**Comment by janryWang** [comment]
给个 codesandbox 看看？

**Reply by RihuaChen** [reply]
https://codesandbox.io/s/cool-sid-sfgpfk?file=/src/App.tsx

点击 Test , Timeout Test可以看到区别

点击Test
<img width="659" alt="image" src="https://github.com/alibaba/formily/assets/11918443/dc80d743-49a7-4c60-9386-321b1552f0a5">
点击 Timeout Test
<img width="666" alt="image" src="https://github.com/alibaba/formily/assets/11918443/9bc2623b-8f09-44ad-8163-98daed8a44c5">

还发现发现如果set过initialValues是OK的，不过如果初始值不设置就不行，辛苦看下

**Comment by janryWang** [comment]
这个不是 bug ，因为生成 graph 树是依赖渲染的，然后渲染过程其实是个异步过程，所以就是这样表现

**Reply by RihuaChen** [reply]
好的好的，那想问一下有办法可以拿到同步的graph吗，或者能不能await这个formGraph呢

**Reply by janryWang** [reply]
这个不能哦，因为渲染时机取决于用户，考虑会存在一些联动
